lib/Sema/MiscDiagnostics.cpp

```diff
@@ -5053,31 +5054,43 @@ static void maybeDiagnoseCallToKeyValueObserveMethod(const Expr *E,
       auto fn = expr->getCalledValue();
       if (!fn)
         return;
-      if (fn->getModuleContext()->getName() != C.Id_Foundation)
-        return;
-      if (!fn->getName().isCompoundName("observe",
-                                        {"", "options", "changeHandler"}))
-        return;
+      SmallVector<KeyPathExpr *, 1> keyPathArgs;
       auto *args = expr->getArgs();
-      auto firstArg = dyn_cast<KeyPathExpr>(args->getExpr(0));
-      if (!firstArg)
-        return;
-      auto lastComponent = firstArg->getComponents().back();
-      if (lastComponent.getKind() != KeyPathExpr::Component::Kind::Property)
-        return;
-      auto property = lastComponent.getDeclRef().getDecl();
-      if (!property)
-        return;
-      auto propertyVar = cast<VarDecl>(property);
-      if (propertyVar->shouldUseObjCDispatch() ||
-          (propertyVar->isObjC() &&
-           propertyVar->getParsedAccessor(AccessorKind::Set)))
+      if (fn->getModuleContext()->getName() == C.Id_Foundation &&
+          fn->getName().isCompoundName("observe",
+                                       {"", "options", "changeHandler"})) {
+        if (auto keyPathArg = dyn_cast<KeyPathExpr>(args->getExpr(0))) {
+          keyPathArgs.push_back(keyPathArg);
+        }
+      } else if (fn->getAttrs()
+                 .hasSemanticsAttr(semantics::KEYPATH_MUST_BE_VALID_FOR_KVO)) {
+        for (const auto& arg: *args) {
+          if (auto keyPathArg = dyn_cast<KeyPathExpr>(arg.getExpr())) {
```

Comment:
So the check only works if, at the call site ,there is a key path literal but it won't work if there another declaration that has a key path type, is that expected from attribute semantics perspective?

Reply:
It's consistent with the existing behaviour applied to the Foundation method.

I'm not sure what you mean by "another declaration that has a key path type". Can you clarify?

Reply:
```
  struct Test {
    var v: Int
  }

  func fn1() -> KeyPath<Test, Int> { \.v }

  //@_semantics("keypath.mustBeValidForKVO")
  func fn2<T, V>(_: KeyPath<T, V>) {}

  fn2(fn1())
```

This would not be diagnosed, I'm just wondering whether this is something which needs to be warned about because key path encodes some of the information that could be used for that...

Reply:
Ah, I see. If the compiler can see that `fn1()` returns an affected key path, then it _should_ be able to emit a diagnostic here already. If it can't see what `fn1()` is returning, then no diagnostic will be emitted and there's really no way around that because there's no way to tell the compiler the `KeyPath` value really is acceptable here.

In any event, the behaviour for `@_semantics()` is consistent with that for the Foundation method, and I'm not looking to change it.